### DISCUSSION TITLE
纯Lua项目如何正确使用xmake

### DISCUSSION THREAD

**Discussion by aceyin** [xmake-io/xmake#2409] [Q&A]
## 背景
我们有一个Lua项目，里面会用到不少C/C++的三方库，也有一些自己写的C/C++库。
我想在这个项目里面使用xmake来管理这些依赖，避免每次升级依赖库的时候，都需要一堆手工操作。

## 问题
我在这个Lua项目里面，写了一个 xmake.lua, 其内容大致如下：

```lua
set_project("a_pure_lua_project")
add_repositories("private_repo http://some-git-url")
add_requires("luasql 2.6.0")
add_requires("sqlite3")
add_requires("lfs")

target("build")
    add_packages("luasql", "sqlite3", "lfs")

    on_build(function(pkg)
        -- 注: 必须重载 on_build 函数，否则在 xmake build 的时候，会报错:  g++: fatal error: no input files
        print("on_build called")
    end)
  
    on_install(function(pkg) 
        print("on_install called")
    end)
```

按照以上方式写好 xmake.lua 之后，问题来了：
因为我重载了 on_build()， 所以导致我的依赖包 (luasql, sqlite3, lfs) 没有被编译
从而导致我最初希望使用 xmake 来编译 .so 文件的目的没法达到

## 求助
所以，想请教一下，如果是我这个情况，应该怎么写这个 xmake.lua， 才能让这些3方依赖库能正常编译出 .so 文件？

注：没有用 luarocks，因为我们有私有库，需要有私有仓库

**Comment by waruqi** [comment]
。。。一个cpp源码都没，还重写了 build 。。这还想要 xmake 干啥，不就是空跑么。。

要编译项目，好歹有点代码吧

**Reply by aceyin** [reply]
> 。。。一个cpp源码都没，还重写了 build 。。这还想要 xmake 干啥，不就是空跑么。。
> 
> 要编译项目，好歹有点代码吧

有代码呀，只不过代码都是Lua而已 :)

其实我是想在 on_build() 里面，调用 xmake 的 dependency 模块，去手工拉去和编译依赖。
但还没研究出来怎么玩

**Reply by waruqi** [reply]
lua 只是脚本，不会参与编译，xmake 主要用于编译 native 代码。。

你只放了 lua ，又想编译出 so 怎么可能。。 so 库都在 add_requires 依赖包里。。

**Reply by aceyin** [reply]
> 

额，我知道作为作者，你一定是惊呆了：“没想到有人会这么用”。
但现实就是这么残酷：偌大一个互联网，连一个 带私有仓库的 Lua 构建工具都找不到。
所以我来xmake这里瞧瞧，主要看中：
1. 搭私有仓库很方便
2. 编译C/C++很简单

但现在就卡在Lua这里。

其实按照我的理解，在纯 Lua 项目里面，去编译 C/C++的依赖，并非不可能。
因为编译依赖和编译项目自身，本身就是一个分离的过程。

我先摸索一下吧

**Reply by aceyin** [reply]
> 你只放了 lua ，又想编译出 so 怎么可能。。 so 库都在 add_requires 依赖包里。。

其实我不是想编译出 Lua 项目的 so 文件啊。
我只想借用xmake的依赖管理功能，能编译出依赖库的 so 文件
